clib/cList.ml

```diff
@@ -241,214 +269,308 @@ let concat l =
 
 let flatten = concat
 
-let rec split_loop p q = function
-| [] -> ()
-| (x, y) :: l ->
-  let cl = { head = x; tail = [] } in
-  let cr = { head = y; tail = [] } in
-  p.tail <- cast cl;
-  q.tail <- cast cr;
-  split_loop cl cr l
+(** {6 Lists as arrays} *)
 
-let split = function
-| [] -> [], []
-| (x, y) :: l ->
-  let cl = { head = x; tail = [] } in
-  let cr = { head = y; tail = [] } in
-  split_loop cl cr l;
-  (cast cl, cast cr)
-
-let rec combine_loop p l1 l2 = match l1, l2 with
-| [], [] -> ()
-| x :: l1, y :: l2 ->
-  let c = { head = (x, y); tail = [] } in
-  p.tail <- cast c;
-  combine_loop c l1 l2
-| _ -> invalid_arg "List.combine"
+let assign l n e =
+  let rec assrec stk l i = match l, i with
+    | (h :: t, 0) -> List.rev_append stk (e :: t)
+    | (h :: t, n) -> assrec (h :: stk) t (pred n)
+    | ([], _) -> failwith "List.assign"
+  in
+  assrec [] l n
 
-let combine l1 l2 = match l1, l2 with
-| [], [] -> []
-| x :: l1, y :: l2 ->
-  let c = { head = (x, y); tail = [] } in
-  combine_loop c l1 l2;
-  cast c
-| _ -> invalid_arg "List.combine"
+(** {6 Filtering} *)
 
 let rec filter_loop f p = function
-| [] -> ()
-| x :: l ->
-  if f x then
-    let c = { head = x; tail = [] } in
-    let () = p.tail <- cast c in
-    filter_loop f c l
-  else
-    filter_loop f p l
+  | [] -> ()
+  | x :: l' as l ->
+    let b = f x in
+    filter_loop f p l';
+    if b then if p.tail == l' then p.tail <- l else p.tail <- x :: p.tail
```

Comment:
This made filter non tailrec

Reply:
Do we have clear guidelines about tail recursivity in our libs?

Reply:
IDK but it seems useless to use the mutation _loop stuff when we're not tailrec.